### PR TITLE
feat: session keys service

### LINKED ISSUE #266: feat: implement session key generation on coordinator service
Implement this https://github.com/privacy-scaling-explorations/maci/issues/1630#issuecomment-2294199993 flow to generate a session key and get it validated by a smart wallet signer

### PR DESCRIPTION
**Pull request by ctrlc03**
Implement a service to handle session keys generation and management 

re: #266

### CODE REVIEW

File: packages/coordinator/package.json [hunk 1]
```diff
@@ -37,30 +37,38 @@
...
+    "@zerodev/permissions": "^5.4.3",
+    "@zerodev/sdk": "^5.3.8",
+    "@zerodev/session-key": "^5.4.2",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.1",
     "dotenv": "^16.4.5",
     "ethers": "^6.13.1",
     "hardhat": "^2.22.6",
     "helmet": "^7.1.0",
+    "lowdb": "^1.0.0",
```
Comment (0xmad) [inline packages/coordinator/package.json:50]: If coordinator supports esm, you can use latest version of lowdb

File: packages/coordinator/ts/file/file.service.ts [hunk 2]
```diff
@@ -19,11 +27,46 @@ export class FileService {
...
+    this.db = low(new FileSync<TStorage>(path.resolve(__dirname, "..", "..", "./session-keys.json")));
+  }
+
+  /**
+   * Store session key
+   *
+   * @param sessionKey - session key
+   * @param address - key address
+   */
+  storeSessionKey(sessionKey: `0x${string}`, address: string): void {
```
Comment (0xmad) [inline packages/coordinator/ts/file/file.service.ts:49]: Let's move ``0x${string}`` to a type

File: packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import dotenv from "dotenv";
+import { ZeroAddress } from "ethers";
+
+import { CryptoService } from "../../crypto/crypto.service";
+import { FileService } from "../../file/file.service";
+import { SessionKeysService } from "../sessionKeys.service";
+
+dotenv.config();
+
+describe("DeployerService", () => {
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts:10]:
```suggestion
describe("SessionKeysService", () => {
```

File: packages/coordinator/ts/sessionKeys/__tests__/utils.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    kernelVersion,
+    plugins: {
+      sudo: ecdsaValidator,
+      regular: permissionPlugin,
+    },
+  });
+
+  const approval = await serializePermissionAccount(sessionKeyAccount);
+
+  return approval;
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/__tests__/utils.ts:81]:
```suggestion
  return serializePermissionAccount(sessionKeyAccount);
```

File: packages/coordinator/ts/sessionKeys/sessionKeys.controller.ts
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/sessionKeys.controller.ts:1]: Let's add tests for controller as well

File: packages/coordinator/ts/file/file.service.ts
Comment (0xmad) [inline packages/coordinator/ts/file/file.service.ts:1]: Let's add tests for added methods

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    const sessionPrivateKey = generatePrivateKey();
+
+    const sessionKeySigner = toECDSASigner({
+      signer: privateKeyToAccount(sessionPrivateKey),
+    });
+
+    const sessionKeyAddress = sessionKeySigner.account.address;
+
+    // save the key
+    this.fileService.storeSessionKey(sessionPrivateKey, sessionKeyAddress);
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:55]: Maybe it makes sense to use redis instead of saving keys to file?

**Review by 0xmad — CHANGES_REQUESTED** [review]
@ctrlc03 thanks, there are some comments

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+@Injectable()
+export class SessionKeysService {
+  /**
+   * Logger
+   */
+  private readonly logger: Logger;
+
+  /**
+   * Create a new instance of DeployerService
+   * @param fileService
```
Comment (kittybest) [inline packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:30]: do we need to add a param for `cryptoService`?

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    return {
+      sessionKeyAddress,
+    };
+  }
+
+  /**
+   * Generate a KernelClient from a session key and an approval
+   *
+   * @param sessionKeyAddress - the address of the session key
+   * @param approval - the approval string
```
Comment (kittybest) [inline packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:66]: --> `encryptedApproval` ?

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts (hunk 5 above)
Comment (ctrlc03) [reply packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:55]: maybe as a future improvement?

Comment (0xmad) [reply packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:55]: Sure, let's do it next. It shouldn't take much time.

File: packages/coordinator/package.json (hunk 1 above)
Comment (ctrlc03) [reply packages/coordinator/package.json:50]: will keep as is if we move to redis in another pr

File: packages/coordinator/ts/sessionKeys/__tests__/utils.ts (hunk 4 above)
Comment (0xmad) [reply packages/coordinator/ts/sessionKeys/__tests__/utils.ts:83]: I think this change is missing

File: packages/coordinator/ts/sessionKeys/sessionKeys.controller.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+   * Delete a session key api method
+   *
+   * @param args - delete session key dto
+   * @returns deleted session key address
+   */
+  @ApiResponse({ status: HttpStatus.CREATED, description: "The session key was successfully deactivated" })
+  @ApiResponse({ status: HttpStatus.FORBIDDEN, description: "Forbidden" })
+  @ApiResponse({ status: HttpStatus.BAD_REQUEST, description: "BadRequest" })
+  @Delete(":sessionKeyAddress")
+  deactivateSessionKey(@Param("sessionKeyAddress") sessionKeyAddress: Hex): void {
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/sessionKeys.controller.ts:48]: Have you tried to send non-hex values? Does it handle validation?

File: packages/coordinator/ts/sessionKeys/__tests__/utils.ts (hunk 4 above)
Comment (0xmad) [reply packages/coordinator/ts/sessionKeys/__tests__/utils.ts:83]: And let's add this file to coverage ignore

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+  /**
+   * Generate a KernelClient from a session key and an approval
+   *
+   * @param sessionKeyAddress - the address of the session key
+   * @param encryptedApproval - the encrypted approval string
+   * @param chain - the chain to use
+   * @returns
+   */
+  async generateClientFromSessionKey(
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:71]: Let's add tests for this function as well.

File: packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      const sessionKeyAddress = sessionKeysService.generateSessionKey();
+      const approval = await mockSessionKeyApproval(sessionKeyAddress.sessionKeyAddress);
+      const encryptedApproval = cryptoService.encrypt(publicKey, approval);
+      await expect(
+        sessionKeysService.generateClientFromSessionKey(
+          sessionKeyAddress.sessionKeyAddress,
+          encryptedApproval,
+          optimismSepolia,
+        ),
+      ).rejects.toThrow();
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts:67]: Let's specify the expected error here.

File: packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    test("should throw when given a non existent session key address", async () => {
+      const approval = await mockSessionKeyApproval(zeroAddress);
+      const encryptedApproval = cryptoService.encrypt(publicKey, approval);
+      await expect(
+        sessionKeysService.generateClientFromSessionKey(zeroAddress, encryptedApproval, optimismSepolia),
+      ).rejects.toThrow();
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/__tests__/sessionKeys.service.test.ts:75]: And the same here

File: packages/coordinator/ts/sessionKeys/sessionKeys.service.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        KERNEL_V3_1,
+        approval,
+        sessionKeySigner,
+      );
+
+      return createKernelAccountClient({
+        bundlerTransport: http(bundlerUrl),
+        entryPoint: ENTRYPOINT_ADDRESS_V07,
+        account: sessionKeyAccount,
+      });
```
Comment (0xmad) [inline packages/coordinator/ts/sessionKeys/sessionKeys.service.ts:120]: Can you add test for success case as well? It's still not covered.

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks!

**Pull request merged**